numpy/core/src/umath/ufunc_object.c

```diff
@@ -998,10 +998,6 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
     }
     if (*allow_legacy_promotion && (!all_scalar && any_scalar)) {
         *force_legacy_promotion = should_use_min_scalar(nin, out_op, 0, NULL);
-        /*
-         * TODO: if this is False, we end up in a "very slow" path that should
-         *       be avoided.  This makes `int_arr + 0.` ~40% slower.
-         */
```

Comment:
I think timings that led me to add it were because of the bad caching.  There is no significant difference to before after this.  (scalars are obviously still slow and do have the added "legacy fallback" overhead.  But the slowdown due to them requiring coercion to an array is the huge chunk.)

If anyone is curious:  I compared 1.21 to the new code again, and some promotion paths do show a small slowdown.  However, I do not think this is due to promotion, but rather due to the casting setup/checking being slightly slower.